### PR TITLE
Fix infinite loop when parsing invalid regular expressions

### PR DESCRIPTION
**Pull request by rozza**
JAVA-2228

### CODE REVIEW

**Comment by jyemin** [comment]
This looks good, but I'm concerned other bugs are hiding.  I don't see, for example, any tests of the escape sequence path, where there is also no check for EOF.

Also, it's worth putting an EOF check in JsonBuffer to avoid infinite loops, rather than continuing to return -1 over and over.

**Comment by rozza** [comment]
>  I don't see, for example, any tests of the escape sequence path, where there is also no check for EOF.

Can you elaborate? I'm not sure what to do here. I checked the other `while(true)` loops and they all handled EOF.  I'll add an EOF check to JsonBuffer as a precaution.

File: bson/src/main/org/bson/json/JsonBuffer.java [hunk 1]
```diff
@@ -34,11 +35,19 @@ public void setPosition(final int position) {
     }
 
     public int read() {
-        return (position >= buffer.length()) ? -1 : buffer.charAt(position++);
+        if (eof) {
+            throw new JsonParseException("Trying to read past EOF.");
+    } else if (position >= buffer.length()) {
```
Comment (jyemin) [inline bson/src/main/org/bson/json/JsonBuffer.java]: Nit: indentation

**Comment by jyemin** [comment]
Sorry, I mis-read the escape sequence code path.  It's fine.

LGTM

File: bson/src/main/org/bson/json/JsonBuffer.java (hunk 1 above)
Comment (rozza) [inline bson/src/main/org/bson/json/JsonBuffer.java]: Done.

**Comment by rozza** [comment]
rebased and merged.

**Comment by rozza** [comment]
Had to revert as I failed to check the other test cases.  

Updated a test to not try and read past eof.
When unreading we now always set eof to false
Updated the JsonReader to not try to read past eof.

**Comment by rozza** [comment]
Also do you want me to backport to 2.14.x ?

File: bson/src/main/org/bson/json/JsonBuffer.java (hunk 1 above)
Comment (jyemin) [inline bson/src/main/org/bson/json/JsonBuffer.java]: Indentation still looks like it's of

File: bson/src/main/org/bson/json/JsonReader.java [hunk 2]
```diff
@@ -226,6 +226,8 @@ public BsonType readBsonType() {
         }
         if (noValueFound) {
             throw new JsonParseException("JSON reader was expecting a value but found '%s'.", token.getValue());
+        } else if (token.getType() == JsonTokenType.END_OF_FILE) {
```
Comment (jyemin) [inline bson/src/main/org/bson/json/JsonReader.java]: This seems to make a bad situation worse.  I see that this allows JsonReaderTest#testEndOfFile1 to pass, but is there any good reason why it _should_ pass?  The test subject is

    "{ test : "

so why should JsonReader generate BsonType.END_OF_DOCUMENT when there is no closing curly brace?

Comment (rozza) [inline bson/src/main/org/bson/json/JsonReader.java]: Fixed

**Comment by jyemin** [comment]
lgtm

**Comment by rozza** [comment]
rebased and merged.

**Pull request closed without merging**